connector/proto/src/main/scala/org/apache/spark/sql/proto/ProtoDataToCatalyst.scala

```diff
@@ -0,0 +1,144 @@
+
+private[proto] case class ProtoDataToCatalyst(child: Expression, descFilePath: String,
+                                              messageName: String,
+                                              options: Map[String, String])
+  extends UnaryExpression with ExpectsInputTypes {
+
+  override def inputTypes: Seq[AbstractDataType] = Seq(BinaryType)
+
+  override lazy val dataType: DataType = {
+    val dt = SchemaConverters.toSqlType(expectedSchema).dataType
+    parseMode match {
+      // With PermissiveMode, the output Catalyst row might contain columns of null values for
+      // corrupt records, even if some of the columns are not nullable in the user-provided schema.
+      // Therefore we force the schema to be all nullable here.
+      case PermissiveMode => dt.asNullable
+      case _ => dt
+    }
+  }
+
+  override def nullable: Boolean = true
+
+  private lazy val protoOptions = ProtoOptions(options)
+
+  @transient private lazy val descriptor = ProtoUtils.buildDescriptor(descFilePath, messageName)
+
+  @transient private lazy val expectedSchema = protoOptions.schema.getOrElse(descriptor)
+
+  @transient private lazy val deserializer = new ProtoDeserializer(expectedSchema, dataType,
+    protoOptions.datetimeRebaseModeInRead)
+
+  @transient private var result: Any = _
+
+  @transient private lazy val parseMode: ParseMode = {
+    val mode = protoOptions.parseMode
+    if (mode != PermissiveMode && mode != FailFastMode) {
+      throw new AnalysisException(unacceptableModeMessage(mode.name))
+    }
+    mode
+  }
+
+  private def unacceptableModeMessage(name: String): String = {
+    s"from_proto() doesn't support the $name mode. " +
+      s"Acceptable modes are ${PermissiveMode.name} and ${FailFastMode.name}."
+  }
+
+  @transient private lazy val nullResultRow: Any = dataType match {
+    case st: StructType =>
+      val resultRow = new SpecificInternalRow(st.map(_.dataType))
+      for (i <- 0 until st.length) {
+        resultRow.setNullAt(i)
+      }
+      resultRow
+
+    case _ =>
+      null
+  }
+
+  private def handleException(e: Throwable): Any = {
+    parseMode match {
+      case PermissiveMode =>
+        nullResultRow
+      case FailFastMode =>
+        throw new SparkException("Malformed records are detected in record parsing. " +
+          s"Current parse Mode: ${FailFastMode.name}. To process malformed records as null " +
+          "result, try setting the option 'mode' as 'PERMISSIVE'.", e)
+      case _ =>
+        throw new AnalysisException(unacceptableModeMessage(parseMode.name))
+    }
+  }
+
+  override def nullSafeEval(input: Any): Any = {
+    val binary = input.asInstanceOf[Array[Byte]]
+    try {
+      result = DynamicMessage.parseFrom(descriptor, new ByteArrayInputStream(binary))
+      val unknownFields = result.asInstanceOf[DynamicMessage].getUnknownFields
+      if (!unknownFields.asMap().isEmpty) {
+        return handleException(new Throwable("UnknownFields encountered"))
```

Comment:
What is the reason for not allowing Unknown fields? This would break forward compatibility right? E.g. the input might have newer versions of protobuf, but the Spark application might be using older version. That is not an incorrect use. The spark job might not needed newer fields.

Reply:
We wanted to support detecting invalid schema. When a protobuf is deserialized with invalid descriptor using FAIL_SAFE or PERMISSIVE, we wanted to raise an assertion or return a null row. The message was successfully deserialized without any failures even in the case of an invalid schema. For example, if a String was deserialized as Boolean, it would successfully deserialize them without any error. Only way to detect them is using UnknownFields. Even in schema evolution, you may add new fields but cannot change the type of an existing field. We could not find the right API to detect invalid schema vs evolved schema.

Reply:
New fields would be so common that we need to allow this.

One way to detect invalid schema vs newer version: We can check if field id of any of the unknown fields matches any of the known fields. That will catch errors like field 3 being in String in original schema and int in incorrect schema.

Reply:
That's a good idea. We will explore this idea.

Reply:
Great. Please add a unit test for that.